### ISSUE TITLE
Assignment through def statement doesn't work with type narrowing

### ISSUE THREAD

**Issue by jdufresne** [python/mypy#9870]
```
$ python --version
Python 3.9.1
$ mypy --version
mypy 0.790
```

Example code:

```py
from typing import Callable, Optional

def f(func: Optional[Callable[[], None]] = None) -> None:
    if func is None:
        def func() -> None:
            pass

    func()
```

Result:

```
$ mypy ~/test.py
/home/jon/test.py:8: error: "None" not callable
Found 1 error in 1 file (checked 1 source file)
```

Expected: No error. The variable `func` has been assigned a callable.

Possible workaround:

```py
from typing import Callable, Optional

def f(func: Optional[Callable[[], None]] = None) -> None:
    if func is None:
        def _func() -> None:
            pass
        func = _func

    func()
```

**Comment by AlexWaygood** [comment]
Duplicate of #2569
